src/box3d.rs

```diff
@@ -0,0 +1,1041 @@
+// Copyright 2013 The Servo Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::UnknownUnit;
+use length::Length;
+use scale::TypedScale;
+use num::*;
+use point::TypedPoint3D;
+use vector::TypedVector3D;
+use side_offsets::TypedSideOffsets3D;
+use size::TypedSize3D;
+use approxord::{min, max};
+
+use num_traits::NumCast;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use core::borrow::Borrow;
+use core::cmp::PartialOrd;
+use core::fmt;
+use core::hash::{Hash, Hasher};
+use core::ops::{Add, Div, Mul, Sub, Neg};
+
+
+/// A 3d box optionally tagged with a unit. 
+/// 
+/// a is the upper left front corner position.
+/// b is the lower right back corner position.
+/// 
+/// Axis directions: x axis positive going left to right.
+///                  y axis positive going bottom to top.
+///                  z axis positive going from back to front (out of the page).
+/// The axis directions follows those used by Microsoft:
+/// https://docs.microsoft.com/en-us/dotnet/framework/wpf/graphics-multimedia/3-d-transformations-overview
```

Comment:
Let's have the 3d box type use a simple min-max representation where `a` is the minimum and `b` is the maximum (you can even call them `min`and `max`).
It's a ton simpler and removes the need to dictate which way is up etc. We have a few methods in TypedRect which specify "top" etc. and it generated complaints so I'd rather not do more of it when it can be avoided (and this avoids me the hassle of peeling web specs since if any convention needs to be adopted it'll have to be one compatible with that).

I realized I might have induced you towards this path in a previous comment by using the "top-left/bottom-right" terminology representation instead of min-max (that's the same thing with the y-is-down convention in the web/gecko/servo), sorry about that.

Reply:
Yea this makes it much easier to comprehend. Will change it.